lib/lightning/workflows.ex

```diff
@@ -190,4 +190,118 @@ defmodule Lightning.Workflows do
       Repo.update_all(workflow_jobs_query, set: [enabled: false])
     end)
   end
+
+  defp filter_digest(digest) do
+    case digest do
+      :daily ->
+        dynamic([r], fragment("age(now(), ?) <= '1 days'", r.finished_at))
+
+      :weekly ->
+        dynamic([r], fragment("age(now(), ?) <= '7 days'", r.finished_at))
+
+      :monthly ->
+        dynamic([r], fragment("age(now(), ?) <= '30 days'", r.finished_at))
+
+      _ ->
+        true
+    end
+  end
+
+  defp successful_workorders_query(workflow, digest) do
+    from(wo in Lightning.WorkOrder,
+      join: w in assoc(wo, :workflow),
+      as: :workflow,
+      join: att in assoc(wo, :attempts),
+      on: wo.id == att.work_order_id,
+      join: r in assoc(att, :runs),
+      as: :runs,
+      join:
+        run in subquery(
+          from(r in Lightning.Invocation.Run,
+            join: att in assoc(r, :attempts),
+            group_by: [att.id, r.exit_code],
+            where: r.exit_code == 0,
+            where: ^filter_digest(digest),
+            select: %{
+              attempt_id: att.id
+            }
+          )
+        ),
+      on: att.id == run.attempt_id,
+      where: w.id == ^workflow.id,
+      select: count(w.id)
+    )
+  end
+
+  defp rerun_workorders_query(workflow, digest) do
+    from(
+      attempt in Lightning.Attempt,
+      as: :attempts,
+      where:
+        1 <
+          subquery(
+            from(wo in Lightning.WorkOrder,
+              join: w in assoc(wo, :workflow),
+              as: :workflow,
+              join: att in assoc(wo, :attempts),
+              on: wo.id == att.work_order_id,
+              join: r in assoc(att, :runs),
+              as: :runs,
+              join:
+                run in subquery(
+                  from(r in Lightning.Invocation.Run,
+                    join: att in assoc(r, :attempts),
+                    group_by: [att.id, r.exit_code],
+                    where: r.exit_code == 0,
+                    where: ^filter_digest(digest),
+                    select: %{
+                      attempt_id: att.id
+                    }
+                  )
+                ),
+              on: att.id == run.attempt_id,
+              where: w.id == ^workflow.id,
+              where: parent_as(:attempts).work_order_id == wo.id,
+              select: count(wo.id)
+            )
+          ),
+      select: count(attempt.id)
+    )
+  end
+
+  defp failed_workorders_query(workflow, digest) do
```

Comment:
On the face of it these queries look good. We do need to:

1. Move them into `WorkOrder` because they have nothing to do with workflows besides a single where clause.
2. Refactor the common parts into queries.